plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/persistence/ElkGraphXMIHelper.java

```diff
@@ -0,0 +1,71 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Christoph Daniel Schulze - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.elk.core.util.persistence;
+
+import org.eclipse.elk.core.data.LayoutMetaDataService;
+import org.eclipse.elk.core.data.LayoutOptionData;
+import org.eclipse.elk.core.util.IDataObject;
+import org.eclipse.elk.core.util.internal.LayoutOptionProxy;
+import org.eclipse.elk.graph.ElkGraphPackage;
+import org.eclipse.elk.graph.properties.IProperty;
+import org.eclipse.emf.ecore.EDataType;
+import org.eclipse.emf.ecore.EFactory;
+import org.eclipse.emf.ecore.xmi.impl.XMIHelperImpl;
+
+/**
+ * An XMI helper which has special knowledge on how to load {@link IProperty} and {@link IDataObject} instances
+ * while loading ELK graphs. A fresh instance must be used for every loading operation. See {@link ElkGraphResource}
+ * for details.
+ * 
+ * @see ElkGraphResource
+ */
+final class ElkGraphXMIHelper extends XMIHelperImpl {
+    
+    private static final ElkGraphPackage ELK_GRAPH_PACKAGE = ElkGraphPackage.eINSTANCE;
+    
+    /** The property that was most recently created. Used to load the next property value. */
+    private LayoutOptionData mostRecentlyCreatedProperty = null;
+    
+    
+    @Override
+    protected Object createFromString(final EFactory eFactory, final EDataType eDataType, final String value) {
+        // We know how to obtain IProperty and parse option values
+        if (eDataType.equals(ELK_GRAPH_PACKAGE.getIProperty())) {
+            return createIPropertyFromString(value);
+        } else if (eDataType.equals(ELK_GRAPH_PACKAGE.getPropertyValue())) {
+            return createPropertyValueFromString(value);
+        } else {
+            return super.createFromString(eFactory, eDataType, value);
+        }
+    }
+    
+    /**
+     * Asks the layout meta data service to load the property which has the given value as its ID. Also saves it
+     * as the most recently loaded property.
+     */
+    private IProperty<?> createIPropertyFromString(final String value) {
+        mostRecentlyCreatedProperty = LayoutMetaDataService.getInstance().getOptionData(value);
+        return mostRecentlyCreatedProperty;
+    }
+    
+    /**
+     * Uses the most recently loaded property to parse the given value as an instance of the property's proper value
+     * type. If no property has been loaded yet, creates a {@link LayoutOptionProxy}.
+     */
+    private Object createPropertyValueFromString(final String value) {
+        if (mostRecentlyCreatedProperty == null) {
+            return new LayoutOptionProxy(value);
+        } else {
+            return mostRecentlyCreatedProperty.parseValue(value);
```

Comment:
Well spotted. Yes, I thought about that. If people don't mess with the XMI output, this should always be true, but if people switch the attributes of a property, this will break. I didn't implement the solution with proxies because I thought that would be too much of an overhead, but now that you mention it I think I will change the implementation accordingly to make it more robust.